### PR TITLE
Add date range field

### PR DESCRIPTION
**Pull request by tintinthong**
Screenshot 2024-11-15 at 22 38 56

### CODE REVIEW

File: packages/host/app/app.ts [hunk 1]
```diff
@@ -10,6 +10,11 @@ import '@cardstack/boxel-ui/styles/global.css';
 import '@cardstack/boxel-ui/styles/fonts.css';
 import '@cardstack/boxel-ui/styles/variables.css';
 import 'ember-power-select/styles';
+import { registerDateLibrary } from 'ember-power-calendar';
+import DateUtils from 'ember-power-calendar-moment';
+import 'ember-power-calendar/styles';
+
+registerDateLibrary(DateUtils);
```
Comment (habdelra) [inline packages/host/app/app.ts]: do you know if this registration mechanism is fastboot friendly? (otherwise our indexer will choke on it)

**Comment by tintinthong** [comment]
I don't think I want to place the onerous of choosing a date library to the user. So I moved the loading of date library into boxel-ui addon and scope it there. Otherwise, we have arbitrary code in host about ember-power-calendar that needs to be loaded there. 

https://github.com/cardstack/boxel/pull/1794/commits/a0ff9caaee48569cbd9cc7d23811f7ea0429790d

File: packages/experiments-realm/productivity/date-range-field.gts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    end: this.args.model.end,
+  };
+
+  get formatted() {
+    if (!this.range.start && !this.range.end) {
+      return '[no date range]';
+    }
+    let start = this.range.start
+      ? Format.format(this.range.start)
+      : '[no start]';
```
Comment (richardhjtan) [inline packages/experiments-realm/productivity/date-range-field.gts]: Not sure if this is necessary, as we will have a start date whenever the user clicks a date. else will be "no date range"

File: packages/experiments-realm/productivity/date-range-field.gts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+class Edit extends Component<typeof DateRangeField> {
+  @tracked range: DateRange = {
+    start: this.args.model.start,
+    end: this.args.model.end,
+  };
+
+  get formatted() {
+    if (!this.range.start && !this.range.end) {
+      return '[no date range]';
```
Comment (richardhjtan) [inline packages/experiments-realm/productivity/date-range-field.gts]: Will "Select a date range" looks better?

File: packages/experiments-realm/productivity/date-range-field.gts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    let start = this.range.start
+      ? Format.format(this.range.start)
+      : '[no start]';
+    let end = this.range.end ? Format.format(this.range.end) : '[no end]';
```
Comment (richardhjtan) [inline packages/experiments-realm/productivity/date-range-field.gts]: "Select end date" may be clearer guidance for users to complete the selection

**Review by richardhjtan — APPROVED** [review]

**Comment by richardhjtan** [comment]
The component looks good when displaying in Boxel UI, but in actual implementation, it somehow looks crowded - possibly due to the lack of padding on the datepicker container, font size, etc. 

We can consider to tweak the styling after that
 
Screenshot 2024-11-14 at 7 03 28 PM

**Comment by tintinthong** [comment]
I realise I need a feature to close the modal for users who have no intent of adding a date AFTER they have opened the calendar. I am will add a button called "Cancel". So the pattern should be quite natural, if the modal is closed, the date is updated BUT if Cancelled, the date is not

File: packages/experiments-realm/productivity/task.gts [hunk 5]
```diff
@@ -808,6 +809,7 @@ export class Task extends CardDef {
   @field dueDate = contains(DateField);
```
Comment (burieberry) [inline packages/experiments-realm/productivity/task.gts:999]: We should remove the `dateStarted` and `dueDate` fields and use the new `dateRange` field instead. Also add a view template to the dateRange component and move the UI that is used in the task cards with the calendar there.

**Comment by tintinthong** [comment]
> I realise I need a feature to close the modal for users who have no intent of adding a date AFTER they have opened the calendar. I am will add a button called "Cancel". So the pattern should be quite natural, if the modal is closed, the date is updated BUT if Cancelled, the date is not

I change my mind about this 

there shud be a
- save button: commits instance and selected date display and close calendar
- reset button: commits instance to null and selected date display and don't close calendar
- click outside: doesn't commit instance and selected date display of last commited and close calendar

**Comment by tintinthong** [comment]
@burieberry I think I want to add some design enhancements to the calendar ie 
- make the dropdown smaller 
- make the row highlights rather than vertical 
- investigate usage of types in registerLibrary

If you review this, I'll merge this in first, otherwise, I'll add those features tomorrow

**Comment by tintinthong** [comment]
Most recent design is here @burieberry 

Screenshot 2024-11-18 at 17 54 36

@lucaslyl has a lot of changes so I just left dueDate and dateStarted in so he can handle them when dealing with isolated

**Pull request merged**